### DISCUSSION TITLE
v1.9.0 release - Feedback Wanted!

### DISCUSSION THREAD

**Discussion by samuelcolvin** [pydantic/pydantic#3538] [General]
See [Changelog](https://pydantic-docs.helpmanual.io/changelog/).

Newest pre-release ready for testing is [**v1.9.0a2**](https://pypi.org/project/pydantic/1.9.0a2/)

**Full List of changes**: https://github.com/samuelcolvin/pydantic/compare/v1.8.2...v1.9.0a2

<hr /><em>This discussion was created from the release <a href='https://github.com/samuelcolvin/pydantic/releases/tag/v1.9.0a1'>v1.9.0a1 (2021-12-18)</a>.</em>

---

**Update:** v1.9.0 is now released.

**Comment by samuelcolvin** [comment]
My plan before release:
* [x] merge any [open PRs with the "ready for review" label](https://github.com/samuelcolvin/pydantic/pulls?q=is%3Aopen+is%3Apr+label%3A%22ready+for+review%22) - currently there are 3, all basically ready to merge
* [x] create a pre-release `v1.9a1` and encourage people to test it before release - I guess I'll give people a week?
* [x] merge #3241
* [x] merge #3540
* [x] merge #3201
* [x] merge #3182
* [x] merge #3542 (resolves #3544)
* [x] merge #3543 (resolves #3541)
* [x] merge #3547 (resolves #3546)
* [x] merge #3551
* [x] release **v1.9.0a2**
* [x] merge #3594
* [x] merge #3595 / fix #3583
* [x] release **v1.9**

My aim is to get the new release out this year.

Are there any other PRs you really really want in this release? If so please comment below (we could create another pre-release if there's something critical to include)

Similarly if you find a bug with this release, please create an issue and link to it here.

**Reply by PrettyWood** [reply]
Small remark but IMO https://github.com/samuelcolvin/pydantic/pull/2231 should be a highlight!

**Reply by samuelcolvin** [reply]
done.

**Comment by christianbundy** [comment]
Thanks for pushing this pre-release! I _think_ the only blocker for my projects is https://github.com/samuelcolvin/pydantic/pull/3241 -- without this merged, we have lots of `KeyError` exceptions that weren't raised in v1.8.2.

**Reply by samuelcolvin** [reply]
Okay, I'll look at it. It's not currently assigned to me though.

**Reply by samuelcolvin** [reply]
Done.

**Reply by christianbundy** [reply]
Thanks! Looks great, and resolves the issue.

**Comment by samuelcolvin** [comment]
from @layday on the issue I've now closed:

> These two typing issues would be good to resolve before release:
> 
> * https://github.com/samuelcolvin/pydantic/pull/3540
> * https://github.com/samuelcolvin/pydantic/pull/2721/files#r771881485
> 
> #2650 is a breaking change and should probably be marked as such - custom encoders might not be equipped to handle `BaseModel`s.

**Reply by samuelcolvin** [reply]
Thanks so much for looking into this

* #3540 merged
* issue on #2721 is complicated, not sure if the issue lies in pydantic or pyright, definitely not a blocker
* I'm not sure about #2650, I assume `json_encoders` should include keys related to models, so the issue shouldn't come up, but maybe I'm wrong, @PrettyWood what do you think?

**Reply by layday** [reply]
Sorry, I wasn't clear. `BaseModel.json` takes a custom encoder as an argument. Before 1.9, custom encoders passed to the `json` method would not have had to handle Pydantic models.

**Reply by PrettyWood** [reply]
I dont see the breaking change. It doesn't need to support a BaseModel class, it just can.
Am I missing something? 🤔

**Reply by layday** [reply]
Pydantic's default encoder won't be executed if you pass an encoder to `json`. As of #2650, nested models won't be converted to a dict. Previously, custom encoders were not expected to handle Pydantic models.

```
$ cat test.py
from pydantic import BaseModel

class MyExoticType:
    pass

def custom_encoder(o):
    if isinstance(o, MyExoticType):
        return ...
    raise TypeError('not serialisable')

class Foo(BaseModel):
    pass

class Bar(BaseModel):
    foo: Foo

Bar(foo=Foo()).json(encoder=custom_encoder)
$ python -m pip list | rg pydantic
pydantic          1.8.2
$ python test.py
$ python -m pip install -q --use-pep517 'pydantic @ git+https://github.com/samuelcolvin/pydantic@v1.9.0a1'
$ python -m pip list | rg pydantic
pydantic          1.9.0a1
$ python test.py
Traceback (most recent call last):
  File "/[...]/test.py", line 17, in <module>
    Bar(foo=Foo()).json(encoder=custom_encoder)
  File "/[...]/venv/lib/python3.9/site-packages/pydantic/main.py", line 488, in json
    return self.__config__.json_dumps(data, default=encoder, **dumps_kwargs)
  File "/nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/lib/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/[...]/test.py", line 9, in custom_encoder
    raise TypeError('not serialisable')
TypeError: not serialisable
```

**Reply by PrettyWood** [reply]
Ah you're absolutely right ! Thanks @layday 👍 I opened https://github.com/samuelcolvin/pydantic/pull/3542. Feedback welcome :)
@samuelcolvin FYI

**Reply by samuelcolvin** [reply]
merged.

**Reply by layday** [reply]
https://github.com/microsoft/pyright/issues/2784 for the typing issue.

**Reply by layday** [reply]
HNY :)

**Reply by samuelcolvin** [reply]
:-(

Will look tomorrow.

**Comment by samuelcolvin** [comment]
from @PrettyWood on the issue I've now closed:

> I reckon https://github.com/samuelcolvin/pydantic/pull/3201 and https://github.com/samuelcolvin/pydantic/pull/3182 are important to fix regressions in `master`/`main`.
> The PR on dataclasses could be very nice too since it fixes **a lot** of issues. I'll see what I can do on Monday or Tuesday.
> Computed fields can and should definitely wait for v1.10: it's a big and impactful change that cannot be rushed and needs perf tests

**Reply by samuelcolvin** [reply]
done

**Comment by andriilahuta** [comment]
@samuelcolvin @PrettyWood #3541

**Reply by samuelcolvin** [reply]
Done

**Comment by christianbundy** [comment]
Just reported another small regression that would block a 1.9.0 upgrade: ~https://github.com/samuelcolvin/pydantic/issues/3544~ https://github.com/samuelcolvin/pydantic/pull/3542

**Reply by christianbundy** [reply]
Resolved!

**Comment by s-demp** [comment]
In the documentation it would be worth adding that to use the .env file with in the Settings class you need to include load_dotenv()

**Reply by samuelcolvin** [reply]
Not sure I understand what you mean? Can you create a new issue.

**Reply by daddycocoaman** [reply]
I think @3051502 is referring to the fact that BaseSettings doesn't appear to load the .env configuration file by itself; you have to call load_dotenv() on the .env file first so that BaseSettings can grab the environment variables, otherwise BaseSettings will yell at you. 

https://github.com/samuelcolvin/pydantic/issues/1482
https://github.com/samuelcolvin/pydantic/issues/1368

**Reply by samuelcolvin** [reply]
You don't need to call `load_dotenv()` as explained in those issues.

I'm sure we could improve the documentation on this and other things, but it has nothing to do with this release.

**Comment by christianbundy** [comment]
Thanks for merging all of these fixes. It looks like all of my tests are passing (!), but now when I run mypy on my codebase (using the pydantic plugin) I'm seeing:

```
mypy.ini:12: error: Error importing plugin "pydantic.mypy": cannot import name
'TypeVarLikeType' from 'mypy.types'
```

It looks like https://github.com/samuelcolvin/pydantic/pull/3552 broke compatibility (intentionally?) with mypy 0.910. Is that expected?

**Reply by christianbundy** [reply]
(In my case, I know that I can upgrade Mypy, but I wouldn't have expected that 1.9.0 would have a backward-incompatible change.)

**Reply by christianbundy** [reply]
Hmm, I've upgraded mypy to the latest release (0.930) and I'm seeing:

```
/.venv/src/pydantic/pydantic/env_settings.py:21: error: INTERNAL ERROR -- Please try using mypy master on Github:
```

**Reply by christianbundy** [reply]
I've opened https://github.com/samuelcolvin/pydantic/pull/3572 to test each supported mypy version so that we can catch regressions in the future.

**Reply by PrettyWood** [reply]
I opened #3573 and fixed the usage of `mypy 0.930` with pydantic in https://github.com/samuelcolvin/pydantic/pull/3573/commits/7168f4c23f35b6a3a591e4f3497b4475460eabcc
I decided to remove the compatibility because it's already a big pain to support so many extra dependencies. Supporting many mypy versions can quickly have a big cost that we need to maintain.
For example with the fix for 0.930, 0.910 will also crash for `builtins` instead of `__builtins__` so we need to add even more `if`...`else`... conditions.
Personally I'm really in favor of supporting only the last `mypy` version for the plugin. It's already hard enough to follow up 😬 

@samuelcolvin WDYT? Should we maintain many mypy versions? And btw we should probably add #3573 in 1.9.0

PS: Merry Christmas everyone 🎅

**Reply by christianbundy** [reply]
Hey, merry Christmas! 🎅  For what it's worth, I'm _very_ happy to see older versions of mypy dropped, but IMO that should be a breaking change that's released under a new major version. Another option would be to version the mypy plugin separately from Pydantic, but that seems like lots of overhead just to avoid a major version bump.

Anyway, I'm sure we'll figure out something that works -- hope you all have nice holidays.

**Reply by christianbundy** [reply]
Fixed!

**Comment by zrothberg** [comment]
See #3583 but the #2650 has some unintended side effects that will probably defeat the vast majority of its usability. As well as cause lots of debugging headaches. Just from reading the code I _think_ we can fix it by updating update_forward_refs but I am not 100% sure. 

Also probably worth expanding the testing to cover intra package namespaces.

**Comment by lmmx** [comment]


**Reply by christianbundy** [reply]
> we need to ask a pydantic maintainer when they will issue a release

You are commenting on a thread where the Pydantic maintaining is issuing a new release.

**Reply by lmmx** [reply]
@christianbundy I know, I was using a block quote to bring it to your attention

**Reply by christianbundy** [reply]
If you are still experiencing #3528 in `master` please open a bug report.

**Reply by lmmx** [reply]
Ah OK, yes I still encountered the bug and followed the guidance to leave feedback here. Will do :+1:

**Comment by christianbundy** [comment]
The last couple times I ran my test suites I made comments indicating which errors I hit -- this time I'm making a comment to confirm there are *zero* errors. LGTM, and thanks for all the hard work on this release. 👍

**Reply by samuelcolvin** [reply]
:tada:

**Comment by jooola** [comment]
I am having some trouble with the new Settings nested environment variables sourcing, finding the right environment variable name is not obvious. And maybe one can tell me if my improvement idea is worth it:

Just landed here, so please correct me if anything is wrong.

Currently the `EnvSettingsSource` is guessing some environment variables name by exploding the os.environ key using the delimiter. This means one might not be able to have a `env_nested_delimiter` to `"_"`. This leads to weird environment variables names (eg `PREFIX_FIELD__SUBFIELD`).

I propose to generate each field's `env_names` during the `Config.prepare_field` and recursively go into the `field.sub_fields`. This will allow for schema introspection `MySettings.schema_json(indent=2)` to print the required env_names to the user.

```py
[...]
        @classmethod
        def prepare_field(cls, field: ModelField) -> None:
            if field.sub_fields is not None:
                for sub_field in field.sub_fields:
                    cls.prepare_field(sub_field)

[...]
```

Additionally these env_names could be used to gather the variables from `os.environ`.

I think it is more elegant to build the env_names keys ( `f"{env_prefix}{fiel_name}{env_nested_delimiter}{sub_field_name}"`) instead of exploding to guess the right key.

**Reply by samuelcolvin** [reply]
Please can you create a new issue.

**Reply by jooola** [reply]
Ok, will do

**Comment by havardthom** [comment]
Thanks for this great release and your hard work! 

Just want to mention that this change was breaking for my FastAPI app:

`stop calling parent class root_validator if overridden, #1895 by @PrettyWood`

Since i had a base model and a subclass with root_validator method having same name. Maybe the change should be highlighted.

**Comment by rslinckx** [comment]
Hi!

I'm using pylance + BaseSettings base class to handle environment variables. It appears that the __dataclass_transform__ thing will synthesise  an __init__ method, but the problem is that for BaseSettings, I never pass explicit values in the constructor, instead I rely on them being in the environment, however pyright will complain about missing arguments for required fields.

Here's an example:

```python
from pydantic import BaseSettings
class S(BaseSettings):
    class Config:
        env_prefix = "FOO_"
    test: str
my_s = S() # pyright: Argument missing for parameter "test" 
```

I would expect in the BaseSettings case to not have such error reported (which is probably specific to BaseSettings, using the BaseModel this is an expected error)

**Reply by christianbundy** [reply]
Did this regress between v1.8.2 and v1.9.0?

**Reply by rslinckx** [reply]
I think there was no _dataclass_transform__ decoration before, so pyyright was mostly ignoring the constructor (and not providing anything helpful). So I wouldn't say it's a regression. But now I have wiggly red lines that say something false :)

**Reply by marianhlavac** [reply]
Just a reminder that for some it's not just a wiggly red lines, but a whole failing builds on CI. It seems that no one is giving this any higher priority, but the latest release is unusable in combination with Python 3.10 features.

**Reply by rslinckx** [reply]
As a workaround, I had "# type: ignore" all constructors to avoid being flagged

**Reply by christianbundy** [reply]
Could you link me to the issue you've created to track this bug? I don't see anything in the issue tracker, but maybe I'm not looking in the right place.

**Reply by marianhlavac** [reply]
There are currently multiple issues opened, which seem to be of the same issue related to the dataclass_transform:
* https://github.com/samuelcolvin/pydantic/issues/3767
* https://github.com/samuelcolvin/pydantic/issues/3753
* https://github.com/samuelcolvin/pydantic/issues/3617

**Reply by marianhlavac** [reply]
It's been over a month and this issue is still being ignored. Can we get it in motion somehow? 

I have a library in which I want to use new Python 3.10 features, but pydantic <1.9.0 does not support them. And pydantic >=1.9.0 is unusable for strictly typed projects.

I'd try to fix it, but seeing large amounts of stale PRs dating back to 2019 I'd rather see an interest from maintainers first, to be sure that such fix would get into upstream.

**Reply by christianbundy** [reply]
If you make a pull request to resolve the issues above and add tests, I'll try to help get it merged.

**Reply by squaresmile** [reply]
I posted a workaround for this issue [here](https://github.com/samuelcolvin/pydantic/issues/3753#issuecomment-1060850457). I agree that this isn't quite a regression, more like a new still functionality that's still being developed by pyright and other libraries.

I think the `dataclass_transform` functionality is generally useful and the workaround not too onerous and you get to keep the new type-checking capabilities. Some documentation about the workarounds might be helpful on the website.

**Reply by samuelcolvin** [reply]
feedback welcome on #3972
